python/cudf/cudf/dataframe/buffer.py

```diff
@@ -135,10 +135,13 @@ def extend(self, array):
     def astype(self, dtype):
         if self.dtype == dtype:
             return self
-        elif self.dtype != "int64" and np.issubdtype(dtype, np.datetime64):
-            return self.astype("int64").astype(dtype)
-        else:
-            return Buffer(cudautils.astype(self.mem, dtype=dtype))
+        elif np.issubdtype(dtype, np.datetime64):
+            numerical_dtype = utils.datetime_to_numerical_dtype(dtype)
+            # if casting to a datetime, we have to cast to the
+            # the corresponding numerical representation first
+            if self.dtype != numerical_dtype:
+                return self.astype(numerical_dtype).astype(dtype)
```

Comment:
How does this handle typecasting between different resolutions? Looks like it will change the actual date erroneously.

Reply:
This is here to handle the case where the source column isn't 64 bit. A real-world scenario might be you have a numeric column of int32s that you want to cast as `datetime64[s]`. In that case we need to cast to int64 first, then view it as `datetime64[s]`.